Userland/Libraries/LibSQL/AST/Insert.cpp

```diff
@@ -37,12 +53,43 @@ RefPtr<SQLResult> Insert::execute(ExecutionContext& context) const
         auto row_value = row_expr.evaluate(context);
         VERIFY(row_value.type() == SQLType::Tuple);
         auto values = row_value.to_vector().value();
-        for (auto ix = 0u; ix < values.size(); ix++) {
-            auto& column_name = m_column_names[ix];
-            row[column_name] = values[ix];
+
+        if (m_column_names.size() > 0) {
+            for (auto ix = 0u; ix < values.size(); ix++) {
+                auto& column_name = m_column_names[ix];
+                auto input_value_type = values[ix].type();
+
+                auto& tuple_descriptor = *row.descriptor();
+                // This must succeed since we checked for every column name for existence in the table.
+                auto element_index = tuple_descriptor.find_if([&](auto element) { return element.name == column_name; }).index();
+                auto element_type = (*row.descriptor())[element_index].type;
+
+                if (!does_value_data_type_match(element_type, input_value_type)) {
+                    return SQLResult::construct(SQLCommand::Insert, SQLErrorCode::InvalidValueType, table_def->columns()[element_index].name());
+                }
+
+                row[element_index] = values[ix];
+            }
+        } else {
+            if (values.size() != row.size()) {
+                return SQLResult::construct(SQLCommand::Insert, SQLErrorCode::InvalidNumberOfValues, "");
+            }
+            for (auto ix = 0u; ix < values.size(); ix++) {
+                auto element_type = (*row.descriptor())[ix].type;
+                auto input_value_type = values[ix].type();
+                if (!does_value_data_type_match(element_type, input_value_type)) {
+                    return SQLResult::construct(SQLCommand::Insert, SQLErrorCode::InvalidValueType, table_def->columns()[ix].name());
+                }
+                row[ix] = values[ix];
+            }
         }
-        context.database->insert(row);
+        inserted_rows.append(row);
+    }
+
+    for (auto& inserted_row : inserted_rows) {
+        context.database->insert(inserted_row);
     }
```

Comment:
Also here remove the braces
```suggestion
    for (auto& inserted_row : inserted_rows)
        context.database->insert(inserted_row);
```